[foothaddeus] iP

foothaddeus opened the pull request:
### Talk-a-Bot

> "Everything begins with an idea." - Earl Nightingale ([source](https://www.brainyquote.com/quotes/earl_nightingale_379872))

Talk-a-bot takes the load off your memory, so you don't have to keep track of your to-dos. It’s:

- text-based
- easy to learn
- ~~SUPER~~ _BLAZING_ FAST to use

Getting started is simple:

1. Download it [here](https://github.com/foothaddeus/ip).
2. Double-click to open.
3. Add your tasks.
4. Let it handle the rest :satisfied:

And the best part? It’s **FREE**!

Features:

- [x] Task management
- [ ] Deadline tracking
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

foothaddeus committed:
Rename, Greet, Exit

foothaddeus committed:
Echo

foothaddeus committed:
Add, List

foothaddeus committed:
Mark as Done (and Undone)

foothaddeus committed:
ToDos, Events, Deadlines

foothaddeus committed:
Automated Text UI Testing

foothaddeus committed:
Handle Errors

Added a few errors to handle including invalid marking and unmarking, no input, unknown input and unknown time

foothaddeus committed:
Add Delete Function

Added the deleting function, and modified the array for storing the to-do list to become an arraylist

foothaddeus committed:
Add the save function

Chatbot only remembers the list while it is running, and memory is lost once the program exits.

Saving the tasks to the hard drive allows the task list to be retrieved even after the program exits.

Add save function, which saves the task list to the hard drive whenever the task list is updated. Also, add load function which checks if task list data is saved onto the hard disk and loads it upon startup, else creates new file.

foothaddeus committed:
Merge branch 'branch-Level-7'

foothaddeus committed:
Add dates

Talk-a-Bot does not interpret deadlines and event timings into dates, and leaves them as a string.

Check if deadline and event dates are valid input, and interprets them as LocalDate instead of string, and allow requests to return the day of the week of these dates.

foothaddeus committed:
Merge branch 'branch-Level-8'

foothaddeus committed:
Use More OOP

Extract Ui, Parser and Tasklist classes to follow the OOP approach more closely.

foothaddeus committed:
Merge branch 'additional-increments'

foothaddeus committed:
Organise into Packages

Divide classes into packages according to their functions for better organisation

foothaddeus committed:
Merge branch 'additional-increments'

foothaddeus committed:
Merge branch 'add-gradle-support'

foothaddeus committed:
Add JUnit Tests

Add tests to test for potential bugs in non-trivial methods such as Event.fileString() and Deadline.getDay()

foothaddeus committed:
Merge branch 'add-junit-tests'

foothaddeus committed:
Merge branch 'master' into additional-increments

foothaddeus committed:
Create a JAR File

Create a JAR File for ease of distribution. Also, fix bugs

foothaddeus committed:
Merge branch 'additional-increments'

foothaddeus committed:
Add JavaDoc Comments

Add JavaDoc comments to public methods and classes for greater clarity on what their functions are.

foothaddeus committed:
Follow the Coding Standard

Tweak code to comply with the coding standard.

foothaddeus committed:
Add Find Function

Add find function for users to find tasks based on a keyword.

foothaddeus committed:
Merge branch 'branch-A-JavaDoc'

foothaddeus committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/talkabot/Parser.java
#	src/main/java/talkabot/Storage.java
#	src/main/java/talkabot/TalkaBot.java

foothaddeus committed:
Merge branch 'branch-Level-9'

jinhanjinhan commented on src/main/java/talkabot/Storage.java:
```diff
@@ -0,0 +1,93 @@
+    private String filePath;
+
+    /**
+     * Constructor for Storage class.
```
Should this be "Constructs a Storage class..." instead? 
```suggestion
     * Constructs a Storage class...

```

jinhanjinhan commented on src/main/java/talkabot/TalkaBot.java:
```diff
@@ -0,0 +1,166 @@
+    private Storage storage;
+
+    /**
+     * Constructor for TalkaBot class.
```
Should this be "Constructs a TalkaBot class..." instead? 
```suggestion
     * Constructs a TalkaBot class...

```

jinhanjinhan commented on src/main/java/talkabot/TalkaBot.java:
```diff
@@ -0,0 +1,166 @@
+    }
+
+    /**
+     * Runs the Talk-a-Bot, responding based on user input.
```
I feel this part can be slightly more specific, like maybe giving an example input and how it works? :)

jinhanjinhan commented on src/main/java/talkabot/TalkaBot.java:
```diff
@@ -0,0 +1,166 @@
+            return false;
+        }
+    }
+
+    private void mark(String input) throws InvalidEditException, IOException {
+        if (!isValidNumber(input, 5)) {
+            throw new InvalidEditException("mark");
+        }
+        Task task = this.taskList.get(Integer.parseInt(input.substring(5)) - 1);
+        task.markAsDone();
+        this.ui.mark(task);
+        storage.save(this.taskList);
+    }
+
+    private void unmark(String input) throws InvalidEditException, IOException {
+        if (!isValidNumber(input, 7)) {
+            throw new InvalidEditException("unmark");
+        }
+        Task task = this.taskList.get(Integer.parseInt(input.substring(7)) - 1);
+        task.markAsUndone();
+        this.ui.unmark(task);
+        storage.save(this.taskList);
+    }
+
+    private void delete(String input) throws InvalidEditException, IOException {
+        if (!isValidNumber(input, 7)) {
+            throw new InvalidEditException("delete");
+        }
+        Task task = this.taskList.delete(Integer.parseInt(input.substring(7)) - 1);
+        this.ui.delete(task, this.taskList.size());
+        storage.save(this.taskList);
+    }
+
+    private void getDay(String input) throws InvalidEditException {
+        if (!isValidNumber(input, 8)) {
+            throw new InvalidEditException("get the day of");
+        }
+        Task task = this.taskList.get(Integer.parseInt(input.substring(8)) - 1);
+        this.ui.getDay(task);
+    }
```
Ahh I see that these are private methods so you did not include Javadoc comments for them. Although not necessary, I thought it would be good if you include some Javadoc comments for these as well so that if you reference to them next time, you will know what the method does easily. Just a thought!

jinhanjinhan commented on src/main/java/talkabot/exceptions/InvalidEditException.java:
```diff
@@ -0,0 +1,14 @@
+public class InvalidEditException extends TalkaBotException {
+
+    /**
+     * Constructor for InvalidEditException class.
```
Should this be "Constructs a InvalidEditException class..." instead? 
```suggestion
     * Constructs a InvalidEditException class...

```

jinhanjinhan commented on src/main/java/talkabot/exceptions/InvalidScheduleException.java:
```diff
@@ -0,0 +1,14 @@
+public class InvalidScheduleException extends TalkaBotException {
+
+    /**
+     * Constructor for InvalidScheduleException class.
```
Same issue for this

jinhanjinhan commented on src/main/java/talkabot/exceptions/NoInputException.java:
```diff
@@ -0,0 +1,15 @@
+public class NoInputException extends TalkaBotException {
+
+    /**
+     * Constructor for NoInputException class.
```
Same constructor issue for this

jinhanjinhan commented on src/main/java/talkabot/exceptions/TalkaBotException.java:
```diff
@@ -0,0 +1,14 @@
+public class TalkaBotException extends RuntimeException {
+
+    /**
+     * Constructor for TalkaBotException class.
```
Same constructor issue for this

jinhanjinhan commented on src/main/java/talkabot/exceptions/UnknownInputException.java:
```diff
@@ -0,0 +1,14 @@
+public class UnknownInputException extends TalkaBotException {
+
+    /**
+     * Constructor for UnknownInputException class.
```
Same constructor issue for this

jinhanjinhan commented on src/main/java/talkabot/exceptions/UnknownTimeException.java:
```diff
@@ -0,0 +1,14 @@
+public class UnknownTimeException extends TalkaBotException {
+
+    /**
+     * Constructor for UnknownTimeException class.
```
Same constructor issue for this

jinhanjinhan commented on src/main/java/talkabot/exceptions/WrongTaskTypeException.java:
```diff
@@ -0,0 +1,14 @@
+public class WrongTaskTypeException extends TalkaBotException {
+
+    /**
+     * Constructor for WrongTaskTypeException class.
```
Same constructor issue for this

jinhanjinhan commented on src/main/java/talkabot/task/Deadline.java:
```diff
@@ -0,0 +1,54 @@
+    private LocalDate deadline;
+
+    /**
+     * Constructor for Deadline class.
```
Same constructor issue for this

jinhanjinhan commented on src/main/java/talkabot/task/Event.java:
```diff
@@ -0,0 +1,56 @@
+    private LocalDate to;
+
+    /**
+     * Constructor for Event class.
```
Same constructor issue for this

jinhanjinhan commented on src/main/java/talkabot/task/Task.java:
```diff
@@ -0,0 +1,63 @@
+    protected boolean isDone;
+
+    /**
+     * Constructor for Task class.
```
Same constructor issue for this, I won't be commenting on this anymore, maybe you can check the other classes regarding this issue as well

jinhanjinhan commented on src/main/java/talkabot/task/Task.java:
```diff
@@ -0,0 +1,63 @@
+        this.isDone = false;
+    }
+
+    /**
+     * Returns status icon of task.
+     *
+     * @return Status icon.
+     */
```
I thought this part can be slightly more specific? Maybe something like giving an example, like e.g. if the task is done, return the status icon X to represent the marking of task. I thought it seems more clear this way when users see the Javadoc. 👍

jinhanjinhan reviewed (commented):
Looks good overall. 👍 👍 
Some minor errors in coding standards, other than that, everything seems great.

foothaddeus committed:
Use CheckStyle

Use CheckStyle to detect coding style violations and resolve them.

foothaddeus committed:
Merge branch 'branch-A-Checkstyle'

nathanshew commented on src/main/java/talkabot/Parser.java:
```diff
@@ -0,0 +1,71 @@
+     * @return String array containing description and deadline.
+     * @throws UnknownTimeException If input does not contain deadline.
+     */
+    public static String[] getDeadline(String input) {
```
I like that you used verbs for methods!

nathanshew commented on src/main/java/talkabot/Storage.java:
```diff
@@ -0,0 +1,91 @@
+     * @return Task list.
+     * @throws IOException When an issue is encountered while trying to create or read file.
+     */
+    public TaskList load() throws IOException {
```
I wonder if it would be better to use loadStorageWithFile() instead?

nathanshew commented on src/main/java/talkabot/TalkaBot.java:
```diff
@@ -0,0 +1,166 @@
+ */
+public class TalkaBot {
+
+    private Ui ui;
```
I wonder if it would be better to use the full form for ui instead of an acronym?

nathanshew commented on src/main/java/talkabot/TalkaBot.java:
```diff
@@ -0,0 +1,166 @@
+    /**
+     * Runs the Talk-a-Bot, responding based on user input.
+     */
+    private void run() {
```
I like that you used a verb for this action!

nathanshew commented on src/main/java/talkabot/TalkaBot.java:
```diff
@@ -0,0 +1,166 @@
+     * Runs the Talk-a-Bot, responding based on user input.
+     */
+    private void run() {
+        boolean end = false;
```
I wonder if isEnd would be better here?

nathanshew reviewed (commented):
Overall well done!

foothaddeus committed:
Add a GUI to the chatbot

Test out creating a GUI, before adding one to the chatbot using the JavaFX technology.

foothaddeus committed:
Merge branch 'branch-Level-10'

foothaddeus committed:
Make minor improvements

Change pictures used, among other improvements

foothaddeus committed:
Use Assertions

No assertions are used in the code.

Assertions can help with testing and debugging.

Use assertions in parts of the code to document important
assumptions that should hold at those points in the code.

This can help with testing debugging as we know what to look for.

foothaddeus committed:
Improve code quality

Methods like run() and getResponse() in the TalkaBot class have weak
SLAP, handling various commands within the method itself, making the
methods long as well.

Follow slap by introducing a new Command class to handle user commands.

This allows run() and getResponse() to maintain a single level of
abstraction.

foothaddeus committed:
Merge pull request #2 from foothaddeus/branch-A-Assertions

Use Assertions

foothaddeus committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/talkabot/TalkaBot.java

foothaddeus committed:
Merge pull request #3 from foothaddeus/branch-A-CodeQuality

Improve code quality

foothaddeus committed:
Make the command syntax more flexible

Talk-a-Bot only interprets commands if they are spelled in full.

Allow shorter command syntax, like "t" for "todo".

Helps users save time by typing shorthands instead of having to key in
the whole command.

foothaddeus committed:
Merge branch 'branch-BCD-Extension'

foothaddeus committed:
Add a representative screenshot and User Guide

Add a screenshot representing the UI of Talk-a-Bot, as well as add a
User Guide for new users to refer to for the features of Talk-a-Bot.

foothaddeus committed:
Merge branch 'branch-A-UserGuide'

foothaddeus committed:
Add Exceptions

Add more Talk-a-Bot exceptions like MissingDetailException and
IncorrectFormatException to help account for more cases of errors
regarding user input.

foothaddeus committed:
Merge branch 'branch-Add-Exceptions'

foothaddeus committed:
Make final changes

Make finishing changes, including editing the behaviour of
MissingDetailException to become MissingDateException.

foothaddeus committed:
Update README.md

foothaddeus committed:
Update README.md
